system/dpfscheduler/pkg/scheduler/cache/blockstate.go

```diff
@@ -156,22 +155,97 @@ func (blockState *BlockState) compute_block_overflow() map[float64]float64 {
+	m := goop.NewModel()
+	// Add your variables to the model
+	x := m.AddBinaryVarVector(len(demands_per_alpha))
+	// Add your constraints
+
+	exp := goop.NewLinearExpr(0)
+	for i := 0; i < len(demands_per_alpha); i++ {
+		exp.Plus(x[i].Mult(demands_per_alpha[i]))
+	}
+	const_expr := goop.NewLinearExpr(a)
+	m.AddConstr(exp.LessEq(const_expr))
+
+	// Set a linear objective using your variables
+	obj := goop.NewLinearExpr(0)
+	for i := 0; i < len(priorities_per_alpha); i++ {
+		obj.Plus(x[i].Mult(float64(priorities_per_alpha[i])))
+	}
+	m.SetObjective(obj, goop.SenseMaximize)
+	// Optimize the variables according to the model
+	sol, err := m.Optimize(solvers.NewGurobiSolver())
+
+	if err != nil {
+		panic("Should not have an error")
+	}
+
+	// Print out the solution
+	//      fmt.Println("x =", sol.Value(x))
+	relval = 0.0
+	for i := 0; i < len(priorities_per_alpha); i++ {
+		if sol.Value(x[i]) > 0 {
+			relval += float64(priorities_per_alpha[i])
+		}
+	}
+	return relval
+}
+
+func (blockState *BlockState) compute_knapsack(claimCache ClaimCache) map[float64]float64 {
+	knapsack_a := map[float64]float64{}
+	demands_per_alpha := map[float64][]float64{}
+	priorities_per_alpha := map[float64][]int32{}
+
+	for claim_id, reservedBudget := range blockState.block.Status.ReservedBudgetMap {
+		reservedBudget.ToRenyi()
+		blockState.block.Status.AvailableBudget.ToRenyi()
+		r, _ := columbiav1.ReduceToSameSupport(reservedBudget.Renyi, blockState.block.Status.AvailableBudget.Renyi)
+		for i := range r {
+			alpha := r[i].Alpha
+			if _, ok := demands_per_alpha[alpha]; !ok {
+				demands_per_alpha[alpha] = make([]float64, 0, 16)
+				priorities_per_alpha[alpha] = make([]int32, 0, 16)
+			}
+			// Fill the array with all the demands for this block/alpha
+			if r[i].Epsilon > 0 {
+				demands_per_alpha[alpha] = append(demands_per_alpha[alpha], r[i].Epsilon)
+				priorities_per_alpha[alpha] = append(priorities_per_alpha[alpha], claimCache.Get(claim_id).claim.Spec.Priority)
+			}
+		}
+	}
+	a := blockState.block.Status.AvailableBudget.Renyi
+	for i := range a {
+		alpha := a[i].Alpha
+		if len(demands_per_alpha[alpha]) > 0 {
+			knapsack_a[alpha] = gurobi_solve(demands_per_alpha[alpha], priorities_per_alpha[alpha], a[i].Epsilon)
+		} else {
+			knapsack_a[alpha] = 0
+		}
+	}
+	return knapsack_a
+}
+
 // UpdateDemandMap updates the demand for each claim on this block
 // - cost
 // - dominant share
 // - availability (enough eps/delta budget, or one alpha positive)
 // - valid
 
-func (blockState *BlockState) UpdateDemandMap() map[string]*DemandState {
+func (blockState *BlockState) UpdateDemandMap(claimCache ClaimCache) map[string]*DemandState {
 	blockState.Lock()
 	defer blockState.Unlock()
 
-	var overflow_a map[float64]float64
-	overflow_a = blockState.compute_block_overflow()
+	var relval map[float64]float64
+	//var knapsack_a map[float64]float64
+	//relval = blockState.compute_block_overflow()
+	relval = blockState.compute_knapsack(claimCache)
```

Comment:
Why do you need to pass the claimCache to compute the knapsack? It seems that you didn't need it for the overflow, and both need the pretty much the same info. Is it because you need to fetch the claim priorities?

Reply:
Exactly!